compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -1694,10 +1694,32 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         // variables
         if substs.has_infer_types_or_consts() {
             let ac = AbstractConst::new(self.tcx, unevaluated.shrink());
-            if let Ok(None) = ac {
-                substs = InternalSubsts::identity_for_item(self.tcx, unevaluated.def.did);
-            } else {
-                return Err(ErrorHandled::TooGeneric);
+            match ac {
+                Ok(None) => {
+                    substs = InternalSubsts::identity_for_item(self.tcx, unevaluated.def.did);
+                    param_env = self.tcx.param_env(unevaluated.def.did);
+                }
+                Ok(Some(ct)) => {
+                    if ct.unify_failure_kind(self.tcx) == FailureKind::Concrete {
+                        substs = self.tcx.mk_substs(substs.iter().enumerate().map(|(idx, arg)| {
+                            if arg.has_param_types_or_consts() || arg.has_infer_types_or_consts() {
+                                // Just using a dummy type here, since the subst doesn't
+                                // actually show up in the abstract const's leafs.
+                                self.tcx
+                                    .mk_ty(ty::Placeholder(ty::PlaceholderType {
+                                        universe: ty::UniverseIndex::MAX,
```

Comment:
you should check here that the generic argument kind is correct, e.g. placeholder type for types and and a placeholder const for constants.

Also, while it doesn't matter, please use the root universe (or the highest universe of all inference variables in that argument) instead of the max universe, as that's closer to what we actually mean.